processor/processor.go

```diff
@@ -1350,18 +1367,68 @@ func (proc *Handle) processJobsForDest(partition string, subJobs subJob, parsedE
 					source,
 				)
 
+				payloadFunc := func() func(json.RawMessage) json.RawMessage {
+					var (
+						called        bool
+						actualPayload json.RawMessage
+					)
+					return func(defaultPayload json.RawMessage) json.RawMessage {
+						if !called {
+							actualPayload = func() json.RawMessage {
+								if proc.transientSources.Apply(source.ID) {
+									return nil
+								}
+								payloadBytes, err := jsonfast.Marshal(singularEvent)
+								if err != nil {
+									return nil
+								}
+								return payloadBytes
+							}()
+							called = true
+						}
+						if actualPayload == nil {
+							return defaultPayload
+						}
+						return actualPayload
+					}
+				}()
```

Comment:
This appears to have too much code. Can we do the following?

1. We create a reusable function in rudder-go-kit
```go
package ro

func Memoize[R any](f func() R) func() R {
	var result R
	var called bool
	return func() R {
		if called {
			return result
		}
		result = f()
		return result
	}
}
```
2. We use it like so
```go
payloadFunc := ro.Memoize(func() json.RawMessage {
	if proc.transientSources.Apply(source.ID) {
		return nil
	}
	payloadBytes, err := jsonfast.Marshal(singularEvent)
	if err != nil {
		return nil
	}
	return payloadBytes
})

...

if payload := payloadFunc(); payload != nil {

...


func() json.RawMessage {
	if p := payloadFunc(); p != nil {
		return p
	}
	return []byte("{}")
},
```

Reply:
[raised a PR here](https://github.com/rudderlabs/rudder-go-kit/pull/30)
Will change here after that goes in.

Reply:
https://github.com/rudderlabs/rudder-go-kit/releases/tag/v0.13.0